### ISSUE TITLE
Performance Of Reasoning Combinators

### ISSUE THREAD

**Issue by oisdk** [agda/agda-stdlib#622]
When I was trying to fix some performance problems which seemed to involve the equational reasoning combinators, I came across [this](https://lists.chalmers.se/pipermail/agda/2016/009090.html) message on the mailing list. It basically suggests replacing the following definition: 

```agda
_∼⟨_⟩_ : ∀ x {y z} → x ∼ y → y IsRelatedTo z → x IsRelatedTo z
_ ∼⟨ x∼y ⟩ relTo y∼z = relTo (trans x∼y y∼z)
```

With this:

```agda
step-∼ : ∀ x {y z} → y IsRelatedTo z → x ∼ y → x IsRelatedTo z
step-∼ _ (relTo y∼z) x∼y = relTo (trans x∼y y∼z)

syntax step-∼ x y∼z x∼y = x ∼⟨ x∼y ⟩ y∼z
```

This changes the order of typechecking, meaning that Agda will look first at `y IsRelatedTo z` to find the definition of `y`, rather than looking at `x ∼ y`. This should yield a pretty big speedup, because `y IsRelatedTo z` will have `y` explicitly stated on the next line (the message mentions a five-fold speedup).

I've found some other benefits, also. The improved type inference means that I can integrate [an automated solver](https://oisdk.github.io/agda-ring-solver/README.html) into the equational reasoning:

```agda
lemma : ∀ x y → x + y * 1 + 3 ≈ 2 + 1 + y + x
lemma x y = begin
  x + y * 1 + 3 ≈⟨ +-comm x (y * 1) ⟨ +-cong ⟩ refl ⟩
  y * 1 + x + 3 ≈⟨ solveOver (x ∷ y ∷ []) Int.ring ⟩
  3 + y + x     ≡⟨⟩
  2 + 1 + y + x ∎
```

I can't get this to work with the current combinators.

How disruptive a change would the flipped combinators be? Would it be worth a pull request?

**Comment by gallais** [comment]
> Would it be worth a pull request?

I think so. The solver integration is a really nice selling point!

**Comment by oisdk** [comment]
Would a new module (Relation.Binary.Reasoning.Flipped or something) be the way to go? Or should I replace the current definitions with the syntax above.

**Comment by gallais** [comment]
I'd say replace them: if I understand nad & you correctly they're basically the same,
but better. Given that we are using the reasoning kit in the stdlib, that will also
give us a nice test case to see whether it breaks existing use cases anywhere.

**Comment by MatthewDaggitt** [comment]
Closed with 2d38f315b6a843f03c694c18c3e789593087ed87
